### ISSUE TITLE
Third party fields integration 

### ISSUE THREAD

**Issue by jonasdumas** [kettanaito/react-advanced-form#126]
I've integrated costom components that works with the React Advanced Form's <Field.x /> component. It works perfectly with all basic forms input. (Input, textarea, select, etc)

But is there any solution to integrate third party fields inside main validation loop?

For example, I'm using a drop files field area and also this following great Select field :
https://github.com/JedWatson/react-select

The solution I see for the moment  is to have another validation loop working in parralal using states in the component where my React Advanced Form (RAF) is. But this become spaghetti where one of the good feature of  RAF is to be the opposite. 

Also, another good point of RAF is to provide a smart solution using the context which let the ability to have uncontrolled components on forms. (Which seems not possible anymore If any special Field with parralal validation)

I would appreciate to have a solution to integrate some field with advanced features inside the main validation loop. Like, for example, a special wrapper <Field.Special /> to wrap a special third party component. This wrapper will let its children component having its own  behavior but with the ability to extract the main value (selected option OR uploaded image) to apply a validation rules to it and treat it as any other RAF Field. (Maybe tru a special callback called onChange/onAnything from the special component)

**Comment by kettanaito** [comment]
There shouldn't be a problem to wrap third party form elements with some wrapper.

In order for this third party element to behave just as native field you would need to map the event handlers (onChange/onFocus/onBlur/etc) on your own.

I think this is possible even now, by extending the `Field` class. This is how all the native fields are implemented - they are extending a generic `Field` class and apply any custom behavior necssary. Roughly, it looks like this:

```
import { Field } from 'react-advanced-form';
import AnyFormElement from 'any-form-lib';

export default class ThirdPartyField extends Field {
  renderElement(props, contextProps) {
    return (<AnyFormElement value={contextProps.get('value')} />);
  }
}
```

This is rather internal functionality now and it is missing from the docs. I was thinking to add it, but then decided it may be unnecessary to know. This example allows you to create completely custom fields which will be registered and will behave as any native field in RAF. However, your example with third party form elements is quite a good one.

Tell me if the example above looks okay to you. I will add this into "Customization" section of the docs.

**Comment by kettanaito** [comment]
@jonasdumas I've added a [Custom fields](https://kettanaito.gitbooks.io/react-advanced-form/customization/custom-fields.html) section into the documentation. This should be what you need to integrate a third-party form elements to work with RAF.

Please let me know if this worked for you. Thanks.

**Comment by jonasdumas** [comment]
Hello @kettanaito 
Thank you so much for your amazing support. 
I've tested your "custom field" solution and it works fine but few remarks that might be interesting for others also:

**1)** I had to extends Field.Input instead of Field:
`export default class SelectField extends Field.Input`

Because this: 
`export default class SelectField extends Field `

Throw the following error:
`TypeError: Super expression must either be null or a function, not object`

Is that a correct implementation ?

**2)** Here's the 'custom field' created. I had to customize the onChange function due to the third-party component requirements (onChange is required with custom output that doesn't return event). I got inspirated from your native handleChange() in Field class. Is that a correct way to register the new value onChange (=whitout "event")?

```
export default class SelectField extends Field.Input {

    handleChange = (selectedValue) => {

      const nextValue = selectedValue;
      const { contextProps } = this;
      const prevValue = contextProps.get('value');

      /* Call parental change handler */
      return this.context.handleFieldChange({
        fieldProps: contextProps,
        nextValue,
        prevValue
      });
    }
    
    renderElement(props, contextProps) {
        return (
            <Select
                {...props}
                value={contextProps.get('value')}
                onChange={this.handleChange}
            />
          );
    }
}
```

**Comment by kettanaito** [comment]
@jonasdumas hi. Thanks for the kind words, means a lot to me!

1) My guess is that `Field` returns an Object, not an instance of internal `Field` class. This is related to how the instances of fields are exported (`Field.Input`,`Field.Select`, etc.). I need to think of the proper way to export either.

Technically, extending any native field to create your custom field should work. However, I don't like the strictness of extension (to create a custom input you need to extend `Field.Input`, to create a custom select - `Field.Select`, when deep down the inheritance/extension chain it doesn't even matter which one you extend, they all inherit from `Field` class), and the semantic problems caused by this. For example, you create a `CustomSelect` but extend a `Field.Input`, which doesn't look and sound natural, making the implementation somewhat ambiguous.

**TL;DR:** Extending any native field to create a custom field should work, but I would recommend, or even enforce, to extend generic `Field` class instead, once I find a way to export it properly.

2) The argument properties you pass to the method handlers is what you will get when providing a custom callbacks to those handlers. If you don't pass `event` reference, then don't expect it to appear in the callback method. Let's assume we are using your `CustomSelect` in some form:

```jsx
export default class MyForm extends React.Component {
  handleSelectChange = ({ event, nextValue, prevValue, fields, form }) => {
    console.log(event); // undefined, since you don't provide it from `CustomSelect.handleChange`
  }

  render() {
    return (
      <Form>
        <CustomSelect name="..." onChange={ this.handleSelectChange } />
      </Form>
    );
  }
}
```

`event` reference is used internally to grab the next value from the form element change. If your form element can provide you the next value without `event` reference - then you can just use it and pass as `nextValue` to `this.context.handleChange`, just as you do. Missing `event` reference is something *you* should keep in mind during the end usage of `CustomSelect`.

Also, note that `props` propagation happens automatically, as I've mentioned this under `Field.renderElement` description in the documentation. You can simplify your `CustomSelect` to this:

```jsx
export default class SelectField extends Field.Input {
    handleChange = (nextValue) => {
      const { contextProps } = this;
      const prevValue = contextProps.get('value');

      /* Call parental change handler */
      return this.context.handleFieldChange({
        fieldProps: contextProps,
        nextValue,
        prevValue
      });
    }
    
    renderElement(props, contextProps) {
        return (
            <Select />
          );
    }
}
```

Three basic event handlers (`onFocus`, `onChange` and `onBlur`) and `value` prop are automatically mapped to the returned React element from `renderElement`. There's nothing wrong to explicitly set them, if it makes it easier to read in our opinion, but I ought to mention it can be simplified.

**Comment by kettanaito** [comment]
@jonasdumas I've also added [Contributing guidelines](https://kettanaito.gitbooks.io/react-advanced-form/developers/contributing.html), so feel free to contribute with your ideas or documentation improvements, if you find some necessary. Each piece of contribution is appreciated.

There isn't going to happen much during the holiday season, but afterward we will try to focus on remaining tasks to reach the 1.0 release. The issues you provide are very useful, as they portray the first impressions of working with RAF. The latter should be outstanding.

**Comment by kettanaito** [comment]
I've got quite a good guide to prefer composition over inheritance, thus the approach mentioned above may not be valid in the next release. I think creation of custom fields will be done via the dedicated HoC.

**Comment by jonasdumas** [comment]
Hello @kettanaito 
Thank you very much for all your answers. I've a more clear idea now about inheritance.
Yes i agree, it would be more clear to be able to directly extends "Field" class directly.

I'm currently stack with the integration of an upload field connected to AWS S3 (not due to RAF).
When this finish i will integrate with RAF as a custom Field.
Do you have an idea for when you will update the way of creating custom fields to fit composition approach ?

**Comment by kettanaito** [comment]
@jonasdumas hi!

I want to merge the proper field compisition this week (maybe even today/tomorrow). I will mention this in here so you would get an update.

**Comment by jonasdumas** [comment]
@kettanaito Thanks for the Contribution Guidelines
Where is the best for you to send proposition for documentation improvements ?
Currently it miss a basic presentation of all available field into Field class

**Comment by kettanaito** [comment]
@jonasdumas feel free to create the issue regarding this, I will add the missing points. I know the list of all available fields is missing, we need to find a good way to illustrate it, since there isn't much to read about each field (they are and behave just as the components they represent - input, select, etc.).
